### ISSUE TITLE
ASN.1 data in PKCS7 signature

### ISSUE THREAD

**Issue by adrienbricchi** [rtyley/spongycastle#15]
Hi,

In a standard PKCS7 signature,  generated with SpongyCastle, in the ASN.1 signature data,
the Sequence objects length seems always set with a 0x80 (undefined) value.

Everything else is correct, though. The length is well set for every other data type.
The structure is identical between Spongy and Bouncy signatures,
but the length isn't undefined with Bouncy (same data to sign and same signature certificate) :
http://img11.hostingpics.net/pics/479305bouncy.png (Boucy)
http://img11.hostingpics.net/pics/696082spongy.png (Spongy)

OpenSSL seems to have hard time to read p7s generated without these length attributes.
The verification is successful using the spongy library, but cannot be validated using some other OpenSSL...

Do you have the same results ? Or any clues ?
Thanks.

**Comment by rtyley** [comment]
I don't have much free time I'm afraid, but what you report is surprising,
given spongy castle is a straight repackage of bouncy castle. I don't cut a
release of SC until I can get the bulk of the test suite to pass (I usually
have to disable some flakey tests, which you can see in the commit history,
but I haven't had to disable any of the ones in this area that I can
recall).

Please do some checks, and report back your results if you can:

What version of BC are you using? Does it match the version of SC? You
should be using BC 1.52 and SC 1.52.X.x

Can you reproduce this issue when running both SC & BC on the regular
OpenJDK VM on your laptop? If not, it could be a problem with the Android
VM. What version of Android and VM are you using? ART or Dalvik?

Can you construct a JUnit test which passes with BC but fails with SC?

**Comment by adrienbricchi** [comment]
Well...

After a few more discussions with the Desktop team in my company, and a proper code digging,
I discovered that the desktop version of our app (that I'm porting to Android) was using Bouncy 1.35 to sign data...
But using sun.security libraries to wrap the signature into a PKCS#7 structure.

The poor ASN.1 structure is exactly the same in Bouncy and Spongy, (tested from 1.35 to 1.52)
The PKCS7 is still valid, though, but not valid-enough for OpenSSL.
After using some sun.security support library, and solving a few bad surprises, I ended up with a perfect PKCS#7 structure.

Sorry for the opened issue, I was told that Bouncy was the only lib used in the other office..
This issue isn't of your concern, it's Bouncy-side, actually.

If anyone has similar issues, feel free to dig up here :
git clone https://adullact.net/anonscm/git/iparapheur-tab/iparapheur-tab.git
My company's projet is opensource (Affero GPL v3).
